### PR TITLE
Add corrections in multi ship record

### PR DESCRIPTION
**Pull request by JohnAFernandez**
The ship record for player ships is not always accurate.  This corrects entries in the ship record as valid packets are received, removing data recorded from guesses while interpolating ship info.

Also, remove a case where null vectors were being passed to the physics engine.

Waiting on some final testing.

Could not have finished this if not for @wookieejedi who has been performing testing and linting until I get a regular setup again.

### CODE REVIEW

**Comment by JohnAFernandez** [comment]
So the main purpose of this is to enhance rollback when dealing when dealing with player ships. How much this increases accuracy is hard to quantify.

File: code/network/multi_interpolate.cpp [hunk 1]
```diff
@@ -89,14 +90,20 @@ void interpolation_manager::interpolate(vec3d* pos, matrix* ori, physics_info* p
...
-		vec3d normalized_rotvel;
-		float mag = vm_vec_copy_normalize(&normalized_rotvel, &pip->rotvel);
-
-		matrix rotate_to_previous;
-
-		vm_quaternion_rotate(&rotate_to_previous, -mag * flFrametime, &normalized_rotvel);
-		vm_matrix_x_matrix(last_orient, &rotate_to_previous, ori);
-
+		if (!IS_VEC_NULL(&pip->rotvel)) {
+			mprintf(("ZZZ TEMP! Confirmed that a vec was *NOT* null A1.\n"));
```
Comment (Goober5000) [inline code/network/multi_interpolate.cpp:94]: Is this ZZZ TEMP warning still supposed to be here?

File: code/network/multi_interpolate.cpp [hunk 2]
```diff
@@ -89,14 +90,20 @@ void interpolation_manager::interpolate(vec3d* pos, matrix* ori, physics_info* p
...
+			vec3d normalized_rotvel;
+			float mag = vm_vec_copy_normalize(&normalized_rotvel, &pip->rotvel);
+
+			matrix rotate_to_previous;
+
+			vm_quaternion_rotate(&rotate_to_previous, -mag * flFrametime, &normalized_rotvel);
+			vm_matrix_x_matrix(last_orient, &rotate_to_previous, ori);
+		} else {
+			mprintf(("ZZZ TEMP! Confirmed that a vec was null A2.\n"));
```
Comment (Goober5000) [inline code/network/multi_interpolate.cpp:103]: Here too

File: code/network/multi_interpolate.cpp [hunk 3]
```diff
@@ -152,21 +159,56 @@ void interpolation_manager::interpolate(vec3d* pos, matrix* ori, physics_info* p
 	vm_interpolate_matrices(ori, &_packets[_prev_packet_index].orientation, &_packets[_upcoming_packet_index].orientation, scale);
 
 	// a quick calculation for the last orientation, courtesy Asteroth
-	vec3d normalized_rotvel;
-	float mag = vm_vec_copy_normalize(&normalized_rotvel, &pip->rotvel);
+	if (!IS_VEC_NULL(&pip->rotvel)) {
+		mprintf(("ZZZ TEMP! Confirmed that a vec was *not* null. B1\n"));
```
Comment (Goober5000) [inline code/network/multi_interpolate.cpp:163]: And here

File: code/network/multi_interpolate.cpp [hunk 4]
```diff
@@ -152,21 +159,56 @@ void interpolation_manager::interpolate(vec3d* pos, matrix* ori, physics_info* p
...
 
-	matrix rotate_to_previous;
+		matrix rotate_to_previous;
 
-	vm_quaternion_rotate(&rotate_to_previous, -mag * flFrametime, &normalized_rotvel);
-	vm_matrix_x_matrix(last_orient, &rotate_to_previous, ori);
+		vm_quaternion_rotate(&rotate_to_previous, -mag * flFrametime, &normalized_rotvel);
+		vm_matrix_x_matrix(last_orient, &rotate_to_previous, ori);
+	} else {
+		mprintf(("ZZZ TEMP! Confirmed that a vec was null. B2\n"));
```
Comment (Goober5000) [inline code/network/multi_interpolate.cpp:172]: And here

File: code/network/multi_obj.cpp [hunk 5]
```diff
@@ -870,6 +870,68 @@ void multi_ship_record_rank_seq_num(object* objp, int seq_num)
...
+
+	if (prev_index < 0 || post_index < 0) { 
+		mprintf(("ZZZ Getting invalid index while trying to update the ship record. If this happens frequently, please investigate!\n"));
+		return;
+	} else if (prev_index == post_index) {
+		mprintf(("ZZZ TEMP! Prev = Post found.\n"));
+		return;
+	}
+
+	mprintf(("ZZZ TEMP! Successully found frame to correct.\n"));
```
Comment (Goober5000) [inline code/network/multi_obj.cpp:912]: And here

File: code/network/multi_obj.cpp [hunk 6]
```diff
@@ -870,6 +870,68 @@ void multi_ship_record_rank_seq_num(object* objp, int seq_num)
...
+
+	auto info = &Oo_info.frame_info[Objects[objnum].net_signature];
+
+	// now that we have valid values, we need to fix the affected values in the record.
+	do {
+
+		mprintf(("ZZZ Temp Positions were %f %f %f\n", info->positions[prev_index].xyz.x, info->positions[prev_index].xyz.y, info->positions[prev_index].xyz.z));
```
Comment (Goober5000) [inline code/network/multi_obj.cpp:919]: And here

File: code/network/multi_obj.cpp [hunk 7]
```diff
@@ -870,6 +870,68 @@ void multi_ship_record_rank_seq_num(object* objp, int seq_num)
...
+		Objects[objnum].interp_info.reinterpolate_previous(
+			Oo_info.timestamps[prev_index], prev_packet_index, current_packet_index,  
+			&info->positions[prev_index], &info->orientations[prev_index], &info->velocities[prev_index], &info->rotational_velocities[prev_index]
+			);
+		++prev_index;
+
+		mprintf(("ZZZ Temp Positions are now %f %f %f\n", info->positions[prev_index].xyz.x, info->positions[prev_index].xyz.y, info->positions[prev_index].xyz.z));
```
Comment (Goober5000) [inline code/network/multi_obj.cpp:926]: And here

**Comment by wookieejedi** [comment]
These print statements were for some final testing, which will be removed. Also, there is an additional fix that needs to be added which we also found, and which will be added once Cyborg is able.

**Comment by JohnAFernandez** [comment]
Ok.  It is now ready for actual testing. : )

**Comment by wookieejedi** [comment]
Tested and this multi update makes a substantial improvement in PvP shots, especially with faster primary weapons and while making shots while turning.

**Review by wookieejedi — APPROVED** [review]
I’ve tested this thoroughly and it provides an immense improvement to PvP. Going over the code multiple times also looks good, so it has my approval. Ideally anyone else interested will also have some time to look it over.

**Comment by wookieejedi** [comment]
Given the above successful comments and review, and no current objection, I'm going to merge this.

**Pull request merged**